src/lib.rs

```diff
@@ -186,39 +350,20 @@ impl<'a, T> LocalBorrowingFfiFuture<'a, T> {
         unsafe extern "C" fn poll_fn<F: Future>(
             fut_ptr: *mut (),
             context_ptr: *mut FfiContext,
-        ) -> FfiPoll<F::Output> {
-            static RUST_WAKER_VTABLE: RawWakerVTable = {
-                unsafe fn clone(data: *const ()) -> RawWaker {
-                    let waker = data.cast::<FfiWaker>();
-                    let cloned = ((*waker).vtable.clone)(waker);
-                    RawWaker::new(cloned.cast(), &RUST_WAKER_VTABLE)
-                }
-                unsafe fn wake(data: *const ()) {
-                    let waker = data.cast::<FfiWaker>();
-                    ((*waker).vtable.wake)(waker);
-                }
-                unsafe fn wake_by_ref(data: *const ()) {
-                    let waker = data.cast::<FfiWaker>();
-                    ((*waker).vtable.wake_by_ref)(waker);
-                }
-                unsafe fn drop(data: *const ()) {
-                    let waker = data.cast::<FfiWaker>();
-                    ((*waker).vtable.drop)(waker);
-                }
-                RawWakerVTable::new(clone, wake, wake_by_ref, drop)
-            };
-
-            // The `waker_ref` is borrowed from external context. We must not call drop on it.
-            let waker = ManuallyDrop::new(Waker::from_raw(RawWaker::new(
-                (*context_ptr).waker_ref.cast(),
-                &RUST_WAKER_VTABLE,
-            )));
+        ) -> MaybeUnwinded<FfiPoll<F::Output>> {
             let fut_pin = Pin::new_unchecked(&mut *fut_ptr.cast::<F>());
-            let mut ctx = Context::from_waker(&*waker);
-            match F::poll(fut_pin, &mut ctx) {
-                Poll::Ready(v) => FfiPoll::Ready(v),
-                Poll::Pending => FfiPoll::Pending,
-            }
+            (*context_ptr).with_as_context(|ctx| {
+                // Unwinding across an FFI boundary is UB
+                // https://doc.rust-lang.org/nomicon/ffi.html#ffi-and-panics
+                //
+                // so we catch the unwind, if any occurs, and then panic on the host.
+                match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```

Comment:
We should document this panic-propagation behavior and that we use `catch_unwind` inside.